File: demoandroid/src/main/java/com/andretietz/retroauth/demo/MainActivity.java
package com.andretietz.retroauth.demo;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.Toast;

import com.andretietz.retroauth.AndroidAuthenticationHandler;
import com.andretietz.retroauth.AndroidTokenType;
import com.andretietz.retroauth.AuthAccountManager;
import com.andretietz.retroauth.Retroauth;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;
import timber.log.Timber;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private static final int RC_ACCOUNT_CHOOSER = 123;
    private GithubService service;
    private AuthAccountManager authAccountManager;

    @SuppressWarnings("ConstantConditions")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Timber.plant(new Timber.DebugTree());

        authAccountManager = new AuthAccountManager(getApplication());

        /**
         * Optional: create your own OkHttpClient
         */
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient httpClient = new OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .build();

        ProviderGithub provider = new ProviderGithub();

        /**
         * Create your Retrofit Object using the {@link Retroauth.Builder}
         */
        Retrofit retrofit = new Retroauth.Builder<>(
                AndroidAuthenticationHandler.create(getApplication(), provider,
                        AndroidTokenType.Factory.create(getApplicationContext())))
                .baseUrl("https://api.github.com/")
                .client(httpClient)
                .addConverterFactory(MoshiConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

//        provider.onRetrofitCreated(retrofit);

        /**
         * Create your API Service
         */
        service = retrofit.create(GithubService.class);


        findViewById(R.id.buttonRequestEmail).setOnClickListener(v -> {
            /**
             * Use it!
             */
            service.getEmails()
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            item -> show(item.toString()),
                            Timber::e
                    );
        });


        findViewById(R.id.buttonInvalidateToken).setOnClickListener(v -> {
            Account activeAccount = authAccountManager
                    .getActiveAccount(getString(R.string.com_andretietz_retroauth_authentication_ACCOUNT));
            if (activeAccount != null) {
                // This is for demo purposes only. We're "manually" setting some-invalid-token.
                AccountManager.get(MainActivity.this).setAuthToken(activeAccount,
                        getString(R.string.com_andretietz_retroauth_authentication_TOKEN), "some-invalid-token");
            }
        });
        findViewById(R.id.buttonInvalidateRefreshToken).setOnClickListener(v -> {
            Account activeAccount = authAccountManager
                    .getActiveAccount(getString(R.string.com_andretietz_retroauth_authentication_ACCOUNT));
            if (activeAccount != null) {
                // This is for demo purposes only. We're "manually" setting some-invalid-token. (refresh token)
                AccountManager.get(MainActivity.this)
                        .setAuthToken(activeAccount,
                                String.format("%s_refresh",
                                        getString(R.string.com_andretietz_retroauth_authentication_TOKEN)),
                                "some-invalid-token");
            }
        });

        findViewById(R.id.buttonSwitchAccount).setOnClickListener(v -> {
            // warning can be ignored when using own account type
            Intent intent = authAccountManager.newChooseAccountIntent(
                    getString(R.string.com_andretietz_retroauth_authentication_ACCOUNT)
            );
            startActivityForResult(intent, RC_ACCOUNT_CHOOSER);
        });

        findViewById(R.id.buttonAddAccount).setOnClickListener(v -> authAccountManager.addAccount(
                getString(R.string.com_andretietz_retroauth_authentication_ACCOUNT),
                getString(R.string.com_andretietz_retroauth_authentication_TOKEN)));

        findViewById(R.id.buttonRemoveAccount).setOnClickListener(view ->
                authAccountManager.removeActiveAccount(getString(R.string.com_andretietz_retroauth_authentication_ACCOUNT)));
    }

    private void show(String toShow) {
        Toast.makeText(this, toShow, Toast.LENGTH_SHORT).show();
    }

    private void showError(Throwable error) {
        Log.e(TAG, "An error occured: ", error);
        show(error.toString());
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK) {
            if (requestCode == RC_ACCOUNT_CHOOSER) {
                String accountType = data.getStringExtra(AccountManager.KEY_ACCOUNT_TYPE);
                String accountName = data.getStringExtra(AccountManager.KEY_ACCOUNT_NAME);
                authAccountManager.setActiveAccount(accountType, accountName);
            }
        }
    }
}

File: demoandroid/src/main/java/com/andretietz/retroauth/demo/MainActivity.kt
package com.andretietz.retroauth.demo

import android.os.Build
import android.os.Bundle
import android.webkit.CookieManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.andretietz.retroauth.AndroidCredentialStorage
import com.andretietz.retroauth.AndroidCredentials
import com.andretietz.retroauth.AndroidOwnerStorage
import com.andretietz.retroauth.Callback
import com.andretietz.retroauth.RetroauthAndroid
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.buttonInvalidateToken
import kotlinx.android.synthetic.main.activity_main.buttonLogout
import kotlinx.android.synthetic.main.activity_main.buttonRequestEmail
import kotlinx.android.synthetic.main.activity_main.buttonSwitch
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import timber.log.Timber

class MainActivity : AppCompatActivity() {
  private lateinit var service: FacebookService

  private val ownerManager by lazy { AndroidOwnerStorage(application) }
  private val credentialStorage by lazy { AndroidCredentialStorage(application) }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    Timber.plant(Timber.DebugTree())

    /**
     * Optional: create your own OkHttpClient
     */
    val interceptor = HttpLoggingInterceptor()
    interceptor.level = HttpLoggingInterceptor.Level.BODY
    val httpClient = OkHttpClient.Builder()
      .addInterceptor(interceptor)
      .build()

    val provider = FacebookAuthenticator(application)

    /**
     * Create your Retrofit Object using the [RetroauthAndroid.createBuilder]
     */
    val retrofit = RetroauthAndroid.createBuilder(application, provider)
      .baseUrl("https://graph.facebook.com/")
      .client(httpClient)
      .addConverterFactory(MoshiConverterFactory.create())
      .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
      .build()

    /**
     * Create your API Service
     */
    service = retrofit.create(FacebookService::class.java)

    buttonRequestEmail.setOnClickListener {
      service.getUserDetails()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
          { item -> show(item.toString()) },
          { error -> showError(error) }
        )
    }

    buttonInvalidateToken.setOnClickListener {
      ownerManager.getActiveOwner(provider.ownerType)?.let { account ->
        credentialStorage.getCredentials(account, provider.credentialType, object : Callback<AndroidCredentials> {
          override fun onResult(result: AndroidCredentials) {
            credentialStorage.storeCredentials(
              account,
              provider.credentialType,
              AndroidCredentials("some-invalid-token", result.data))
          }

          override fun onError(error: Throwable) {
            showError(error)
          }
        })
      }
    }

    buttonLogout.setOnClickListener {
      ownerManager.getActiveOwner(provider.ownerType)?.let { account ->
        ownerManager.removeOwner(account.type, account, object : Callback<Boolean> {
          override fun onResult(result: Boolean) {
            show("Logged out: $result")
          }

          override fun onError(error: Throwable) {
            showError(error)
          }
        })
      }
      /** remove all cookies to avoid an automatic relogin */
      val cookieManager = CookieManager.getInstance()
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
        @Suppress("DEPRECATION")
        cookieManager.removeAllCookie()
      } else {
        cookieManager.removeAllCookies(null)
      }
    }

    buttonSwitch.setOnClickListener {
      ownerManager.switchActiveOwner(provider.ownerType)
      /** remove all cookies to avoid an automatic relogin */
      val cookieManager = CookieManager.getInstance()
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
        @Suppress("DEPRECATION")
        cookieManager.removeAllCookie()
      } else {
        cookieManager.removeAllCookies(null)
      }
    }
  }

  private fun show(toShow: String) {
    Toast.makeText(this, toShow, Toast.LENGTH_SHORT).show()
  }

  private fun showError(error: Throwable) {
    show(error.toString())
  }
}
